engine/engine.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package engine
+
+import (
+	"os"
+
+	"github.com/rs/zerolog"
+)
+
+type Engine struct {
+	log        zerolog.Logger
+	components []*Component
+
+	sig    chan os.Signal
+	done   chan struct{}
+	failed chan struct{}
```

Comment:
Actually I changed it to just one single channel of `error` where we send `nil` when it exits successfully without errors and the error otherwise. I think it does what you wanted and it's also simpler I believe :)

Reply:
Yeah, the only reason I suggested a struct is the logging, a plain `error` is more elegant.